Rest client using multipart stopped working while migrating to Quarkus 3

### Discussed in https://github.com/quarkusio/quarkus/discussions/33600

<div type='discussions-op-text'>

<sup>Originally posted by **PierreBtz** May 25, 2023</sup>
Hey folks,

While migrating to Quarkus 3.x, I noticed an issue with one of my RestClients. I managed to reduce this to a minimal example showing the issue, but I cannot determine whether I'm doing something wrong that used to work by chance or if I found a genuine bug. Here goes:

```
@ApplicationScoped
@RegisterRestClient(configKey = "client")
public interface Client {

    @POST
    @Path("/api")
    void call(MultipartBody data);

    class MultipartBody {

        @RestForm("file")
        @PartType(MediaType.APPLICATION_OCTET_STREAM)
        private final InputStream file;

        public MultipartBody(InputStream file) {
            this.file = file;
        }

        public InputStream getFile() {
            return file;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            MultipartBody that = (MultipartBody) o;
            try {
                return IOUtils.contentEquals(file, that.file);
            } catch (IOException e) {
                return false;
            }
        }

        @Override
        public int hashCode() {
            return Objects.hash(file);
        }
    }
}
```
Simply running the following test with Quarkus 3.x:

```
@QuarkusTest
class ClientTest {

    @Inject
    @RestClient
    Client client;

    @Test
    void aTest() {
    }

}
```

Gives:

```
Caused by: java.lang.IllegalArgumentException: Not a REST client interface: interface eu.pierrebeitz.bug.Client. No @Path annotation found on the class or any methods of the interface and no HTTP method annotations (@POST, @PUT, @GET, @HEAD, @DELETE, etc) found on any of the methods
```

Now going back to the latest 2.x release, and adapting the import s/jakarta/javax, I have a working example.

The issue is clearly linked to the usage of the multipart object for some reason, but I don't see where I'm wrong/if I'm wrong. I didn't find any issue/PR related to this either.

Repo in case someone wants to have a look: https://github.com/PierreBtz/quarkus-rest-easy-client-bug

Any ideas?

</div>

--- a/extensions/resteasy-reactive/jaxrs-client-reactive/deployment/src/main/java/io/quarkus/jaxrs/client/reactive/deployment/JaxrsClientReactiveProcessor.java
+++ b/extensions/resteasy-reactive/jaxrs-client-reactive/deployment/src/main/java/io/quarkus/jaxrs/client/reactive/deployment/JaxrsClientReactiveProcessor.java
@@ -1706,10 +1706,12 @@ public class JaxrsClientReactiveProcessor {
 
     private void addInputStream(BytecodeCreator methodCreator, AssignableResultHandle multipartForm, String formParamName,
             String partType, String partFilename, ResultHandle fieldValue, String type) {
+        ResultHandle formParamResult = methodCreator.load(formParamName);
+        ResultHandle partFilenameResult = partFilename == null ? formParamResult : methodCreator.load(partFilename);
         methodCreator.assign(multipartForm,
                 methodCreator.invokeVirtualMethod(MethodDescriptor.ofMethod(QuarkusMultipartForm.class, "entity",
                         QuarkusMultipartForm.class, String.class, String.class, Object.class, String.class, Class.class),
-                        multipartForm, methodCreator.load(formParamName), methodCreator.load(partFilename), fieldValue,
+                        multipartForm, formParamResult, partFilenameResult, fieldValue,
                         methodCreator.load(partType),
                         // FIXME: doesn't support generics
                         methodCreator.loadClassFromTCCL(type)));

--- a/extensions/resteasy-reactive/rest-client-reactive/deployment/src/test/java/io/quarkus/rest/client/reactive/multipart/MultipartFilenameTest.java
+++ b/extensions/resteasy-reactive/rest-client-reactive/deployment/src/test/java/io/quarkus/rest/client/reactive/multipart/MultipartFilenameTest.java
@@ -111,6 +111,10 @@ public class MultipartFilenameTest {
         // Using a field form param
         assertThat(client.postMultipartWithPartFilenameUsingInputStream(new ByteArrayInputStream(bytes)))
                 .isEqualTo(EXPECTED_OUTPUT);
+        // Using rest data annotation without filename
+        ClientRestFormUsingInputStream restForm = new ClientRestFormUsingInputStream();
+        restForm.file = new ByteArrayInputStream(bytes);
+        assertThat(client.postMultipartWithPartFilenameUsingInputStream(restForm)).isEqualTo("myFile:" + FILE_CONTENT);
     }
 
     @Test
@@ -279,6 +283,11 @@ public class MultipartFilenameTest {
         @POST
         @Path("/using-form-data")
         @Consumes(MediaType.MULTIPART_FORM_DATA)
+        String postMultipartWithPartFilenameUsingInputStream(@MultipartForm ClientRestFormUsingInputStream clientForm);
+
+        @POST
+        @Path("/using-form-data")
+        @Consumes(MediaType.MULTIPART_FORM_DATA)
         String postMultipartWithPartFilenameUsingInputStream(
                 @FormParam("myFile") @PartType(APPLICATION_OCTET_STREAM) @PartFilename(FILE_NAME) InputStream file);
 
@@ -342,6 +351,12 @@ public class MultipartFilenameTest {
         @FormParam("myFile")
         @PartType(APPLICATION_OCTET_STREAM)
         @PartFilename(FILE_NAME)
+        public InputStream file;
+    }
+
+    public static class ClientRestFormUsingInputStream {
+        @RestForm("myFile")
+        @PartType(APPLICATION_OCTET_STREAM)
         public InputStream file;
     }
 

